src/ServiceControl.Monitoring/Raw/RawDataProvider.cs

```diff
@@ -1,29 +1,83 @@
+            var endpointName = headers.GetOriginatingEndpoint();
+            var timestamp = data["Timestamp"].Value<DateTime>();
+            var criticalTimeMetrics = data["Timers"].First(elem => elem["Name"].Value<string>() == "Critical Time");
+            var criticalTime = criticalTimeMetrics["Histogram"]["Mean"].Value<float>();
+            var processingTimeMetrics = data["Timers"].First(elem => elem["Name"].Value<string>() == "Processing Time");
+            var processingTime = processingTimeMetrics["Histogram"]["Mean"].Value<float>();
+
+            // ordered collections are not threadsafe
+            var endpointLock = endpointLocks.GetOrAdd(endpointName, new object());
+
+            lock (endpointLock)
+            {
+                if (!criticalTimeValues.ContainsKey(endpointLock))
+                    criticalTimeValues[endpointLock] = new SortedList<DateTime, float>();
+                criticalTimeValues[endpointLock].Add(timestamp, criticalTime);
+
+                if (!processingTimeValues.ContainsKey(endpointLock))
+                    processingTimeValues[endpointLock] = new SortedList<DateTime, float>();
+                processingTimeValues[endpointLock].Add(timestamp, processingTime);
+            }
+        }
+
+        public dynamic GetValues(TimeSpan range)
+        {
+            var historicMax = DateTime.UtcNow.Subtract(range);
+            var result = new List<EndpointMetrics>();
+            foreach (var endpoint in endpointLocks)
+            {
+                var metrics = new EndpointMetrics(endpoint.Key);
+
+                var endpointLock = endpoint.Value;
+                lock (endpointLock)
+                {
+                    // check because of potential race condition
+                    if (criticalTimeValues.ContainsKey(endpointLock))
+                    {
+                        var endpointCriticalTime = criticalTimeValues[endpointLock];
+                        metrics.CriticalTime = endpointCriticalTime.SkipWhile(x => x.Key < historicMax).Select(x => x.Value).ToArray();
+                    }
+
+                    // check because of potential race condition
+                    if (processingTimeValues.ContainsKey(endpointLock))
+                    {
+                        var endpointProcessingTime = processingTimeValues[endpointLock];
+                        metrics.ProcessingTime = endpointProcessingTime.SkipWhile(x => x.Key < historicMax).Select(x => x.Value).ToArray();
+                    }
+                }
+
+                result.Add(metrics);
+            }
+
+            return result;
         }
 
-        ConcurrentDictionary<string, JObject> contexts = new ConcurrentDictionary<string, JObject>();
+        readonly ConcurrentDictionary<string, object> endpointLocks = new ConcurrentDictionary<string, object>();
+        readonly Dictionary<object, SortedList<DateTime, float>> criticalTimeValues = new Dictionary<object, SortedList<DateTime, float>>();
```

Comment:
If we could afford storing k-last entries, instead of a `SortedList<DateTime,float>` one could provide for each endpoint a circullar buffer with an index per endpoint. This would enable lockfree, Interlocked based operations. This might be not needed though (requires knowledge about frequency of operations etc etc)